### ISSUE TITLE
IC/BC caching causes critical errors in inverse problems for IC/BC

### ISSUE THREAD

**Issue by kyouma** [lululxvi/deepxde#2074]
Hello.

I have tried to solve an inverse problem for the same equation as in #2073: `2.4` in the IC is changed into `LAMBDA`, and some data points ("measurements") are connected into the BC list. I haven't use the `anchors` argument, because even without anchors the whole `PointSetBC` data point set is connected into the training and testing sets by DeepXDE:
```
class PointSetBC:
    ...
    def collocation_points(self, X):
        if self.batch_size is not None:
            self.batch_indices = self.batch_sampler.get_next(self.batch_size)
            return self.points[self.batch_indices]
        return self.points
```

I have found out the following error while working with the modified DeepXDE ([this commit](https://github.com/kyouma/deepxde/commit/b0ffb52873203945a09760c540ecdce93620c67a) to fix the points mismatch during test loss calculations), but I think that the same applies to the stock DeepXDE (main branch, 30.03.2026).

I found out that the IC loss value on the test sample was very high (for example, O(1) or O(0.1)) even when the training IC loss value was low enough and the metrics with the ground truth solution were also low. I decided to calculate the IC loss myself manually using `data.train_x` and `data.test_x` points, and the manual calculation result was even lower than that of the training IC loss value in the log.

In ICs and BCs the function `func` is cached using `npfunc_range_autocache()`, and when the trainable variables involved in calculation of this `func` changes, the return value doesn't change, and moreover, the variable gradient becomes incorrect. If I disable the cache logic in `wrapper_cache()` and `wrapper_cache_auxiliary()`, my inverse problem converges easily, while the variable flew to `>10` with the caching enabled.

Is this a known behavior? Is there a way to fix this? Maybe by passing a list of trainable variables involved in the function calculation. Or just disabling the caching? I can try to prepare a pull request if I get some guidance about DeepXDE caching and coding style politics.

I also don't know how Tensorflow handles such situations: I see that DeepXDE doesn't force it to cache the function values.

**Comment by kyouma** [comment]
If I disable the caching in a simple non-inverse problem (_2D space X 1D time; 1 equation with 1st and 2nd derivatives, 4+1 conditions; approx. 5000 points per the domain, boundary and initial condition; no `external_trainable_variables`_), for PyTorch on a RTX 2080 Ti the performance decreases by 15%.

At least one may disable the cache for all BCs and IC only if `external_trainable_variables` is non-empty, and for forward problems nothing changes. If it is OK, I can try to make a pull request.

**Comment by echen5503** [comment]
Could you please clarify a few things about this issue?

1. When you say " for PyTorch on a RTX 2080 Ti the performance decreases by 15%." are you referring to the code slowing down? Speeding up? Or referring to the loss?
2. What does O(1), O(0.1) mean? Do you mean "on the order of?" note that O() is reserved for asymptotic performance, not numerical values.
3. Ensure this is also an issue for deepXDE master. 

You can get some guidance on deepXDE caching in the deepXDE slack, If you are interested I can send you an invite.

**Comment by kyouma** [comment]
I am sorry, I have not received the notification about your reply before my PR.

1. Yes, when I tried a simple inverse problem with 1 PDE, 4 BCs, 1 IC, 1 data `PointSetBC` with caching completely disabled for all BC/IC functions, the performance was 15-17% lower that with caching enabled. The measurements were done for PyTorch backend on a single RTX 2080 Ti.
2. Yes, I meant the order.
3. It is, so I have made a PR.

As for the invite, I appreciate your proposal, but I am not sure that I can contribute to DeepXDE often. If this is not a problem, then it would be great to receive one.

**Comment by echen5503** [comment]
Ok, thanks for the clarification; let's redirect discussion to #2077. 

For the slack, there is no need to be active maintainer, just a moderate/advanced user. If you would like, contact me at [echen1ffa@gmail.com](mailto:echen1ffa@gmail.com) and I'll send you an invite.

<img width="598" height="59" alt="Image" src="https://github.com/user-attachments/assets/4e0d6418-4646-4371-8d84-a2fbdae39957" />
